### ISSUE TITLE
includes in `VC/Tools/MSVC/14.36.32530/include` are not lowercased

### ISSUE THREAD

**Issue by huangqinjin** [mstorsjo/msvc-wine#72]
E.g. `VC/Tools/MSVC/14.36.32530/include/comdef.h` includes `Ole2.h` and `OleCtl.h`.

**Comment by mstorsjo** [comment]
Does https://github.com/mstorsjo/msvc-wine/commit/03fed75207ef14800f1c5994c031f129566cfec2 fix this issue?

But does it cause further issues? These headers are mostly lowercase, so it should be mostly fine. 

A quick check at these headers find the following ones that don't have entirely lowercase names:
```
./Manifest
./Manifest/dpiaware.manifest
./Manifest/PerMonitorHighDPIAware.manifest
./CodeAnalysis
./CodeAnalysis/HResultCheck
./CodeAnalysis/HResultCheck/warnings.h
./CodeAnalysis/ConcurrencyCheck
./CodeAnalysis/ConcurrencyCheck/warnings.h
./CodeAnalysis/Warnings.h
./CodeAnalysis/VariantClear
./CodeAnalysis/VariantClear/warnings.h
./CodeAnalysis/sourceannotations.h
./CodeAnalysis/CppCoreCheck
./CodeAnalysis/CppCoreCheck/warnings.h
./CodeAnalysis/EnumIndex
./CodeAnalysis/EnumIndex/warnings.h
./fuzzer/FuzzedDataProvider.h
```

At least the `CodeAnalysis/sourceannotations.h` header does get included by `sal.h`, so I guess we'd need to lowercase these files too?

**Comment by huangqinjin** [comment]
> Does https://github.com/mstorsjo/msvc-wine/commit/03fed75207ef14800f1c5994c031f129566cfec2 fix this issue?

Yes. It does!

> we'd need to lowercase these files too?

I do think so. We cannot control how users include these headers, but as least we should make headers installed self-consistent. So that issues like `ninja` dependencies and `ccache` cache miss are not due to these installed headers, and users have the chance to fix these issues in  their code.

**Comment by mstorsjo** [comment]
> > Does [03fed75](https://github.com/mstorsjo/msvc-wine/commit/03fed75207ef14800f1c5994c031f129566cfec2) fix this issue?
> 
> Yes. It does!

Actually, since the WinSDK part of the headers are lowercased with replacement symlinks, I wondered why this was needed at all. But I see that these versions of the names, `Ole2.h` and `OleCtl.h`, don't exist in that form at all in the WinSDK. So we do need to do this indeed...

> > we'd need to lowercase these files too?
> 
> I do think so. We cannot control how users include these headers, but as least we should make headers installed self-consistent. So that issues like `ninja` dependencies and `ccache` cache miss are not due to these installed headers, and users have the chance to fix these issues in their code.

Yeah, I guess so.

I pushed an amended version at 377a196847c4b34214008c2ef5ddf3df946c5cd5.

**Comment by huangqinjin** [comment]
LGTM!

**Comment by mstorsjo** [comment]
> In `VC/Tools/MSVC/14.36.32532/include/limits`:
> 
> ```
> #include _STL_INTRIN_HEADER
> ```
> 
> becomes
> 
> ```
> #include _stl_intrin_header
> ```

Oh, tricky. Amended the branch as 51d3380d0e4f78309d4c239bda33fbb546d7cfa9.

**Comment by huangqinjin** [comment]
LGTM now, built OpenCV twice, no issue observed.

**Comment by huangqinjin** [comment]
Maybe we can add a testcase, simply include all commonly used headers in one source file and compile it?

**Comment by mstorsjo** [comment]
> Maybe we can add a testcase, simply include all commonly used headers in one source file and compile it?

I guess that sounds reasonable.

A file that should trigger many of the issues we've observed here should be this:
```
#include <windows.h>
#include <comdef.h>
#define _USE_DECLSPECS_FOR_SAL=0
#define _USE_ATTRIBUTES_FOR_SAL=1 
#include <sal.h>
```
I guess we could list dozens of more individual files too, but I don't know how much more specific test coverage that gives, unless you have other cases in mind?

**Comment by mstorsjo** [comment]
> LGTM now, built OpenCV twice, no issue observed.

Pushed these fixes to master, then.

**Comment by mstorsjo** [comment]
> > we could list dozens of more individual files
> 
> That is what in my mind.

Ok, great. FWIW, a test like this would be good to execute with clang-cl too, to test the headers in case sensitive mode. (We probably don't need to check the deps and build it with ninja, just compiling it manually once should be enough.)

Do you want to add this test into #74?

**Comment by mstorsjo** [comment]
> We still need to check deps with ninja, especially for STL headers.

Yes, that's definitely necessary to test - but for testing the case sensitiveness, a plain compile with clang-cl is good extra coverage too.

> STL uses latest MSVC compiler features which are probably not available in Clang. My thought is to build https://github.com/microsoft/STL/blob/main/stl/modules/std.ixx directly, so that we do not need to track c++ standard headers manually.

Ok - that might be reasonable, I don't quite know at the moment - I have zero experience with C++ modules so far.

> Build Win32 and UCRT headers with clang-cl would be good though. I don't have a list of headers in mind now, so I don't want to add this to #74. The `test-clang-lld` need refactor too and it fails with Clang16 on my PC, I think `-fuse-ld=lld` is needed but no idea why CI passes.

Ok, I can take a stab at that after you otherwise feel #74 is ready to be merged (it's still marked as draft).

You're right that it's weird if that fails for you but passes in CI.

When CMake itself uses `clang-cl`, it will generally default to doing the separate linking step with `lld-link` instead of `link`, so that should generally work. But it's possible that some step tries to do a plain compile+link with one command, and then `clang-cl` will default to trying to find `link.exe`, which might fail. Even if that would find it though, that CI environment lacks Wine so it can't be doing that. So you're right that it's somewhat mysterious.

If you want to dig further into that, a CI run that prints the `CMakeFiles/CMake{Output.log,Error.log,ConfigureLog.yaml}` afterwards, and comparing that to yours, probably could shed some light on that difference.

**Comment by huangqinjin** [comment]
OK, thanks for the explanations. And I find this thread https://discourse.llvm.org/t/how-to-specify-winsysroot-argument-without-using-clang-cl/61643, the `/winsysroot` option isn't supported by `clang`.

**Comment by huangqinjin** [comment]
Closes as #75 landed.
